### ISSUE TITLE
assign_moving_average function explodes when smoothing bounded inputs

### ISSUE THREAD

**Issue by MCMcCallum** [tensorflow/tensorflow#21726]
### System information
- Problem encountered when running custom code
- Linux Ubuntu 16.04 LTS
- TensorFlow installed from source
- TensorFlow version 1.8.0
- Python 3.5.2 
- Bazel Version 0.11.1
- GCC 4.9.4
- CUDA 9.1.85, CuDNN 7.1.2
- GPU: Gtx1070, 8GB memory

### Describe the problem
In training a network consisting of the architecture described below, I encountered a problem with batch normalization layers. I found that the Keras batch normalization moving average parameters were unstable. While decreasing loss was observed over several iterations, it would regularly explode to a very high value. In observing the batch normalization weights in Tensorboard I notice that the moving mean in batch normalization layers explode at the same iteration of this loss explosion, while the batch statistics remain well behaved as shown in the picture below. This doesn't add up as the moving averages are a stable function of the batch statistics.

I noticed that the only Keras code employed in updating these parameters is the `assign_moving_average` function in `tensorflow/tensorflow/python/training/moving_averages.py`. Looking deeper this function in turn uses the `assign_sub` function in `tensorflow/tensorflow/python/ops/state_ops.py` which by default does not use any locking and hence its behaviour may be undefined. Perhaps this is the cause, and if so, should the default be to ignore locking? How certain is it that this wont run into undefined behaviour?

When I replace the Tensorflow `assign_moving_average` function with a simple handcrafted `(1-self.momentum)*mean + (self.momentum)*self.moving_mean` in the Keras BatchNormalization object, this problem never occurs.

### Source code / logs
For reference I am training a three layer convolutional network with three max pooling layers and three batch normalization layers, followed by two dense layers and two batch normalization layers at the output. The network employs a triplet loss and so the weights are each employed three times prior to back propogation. I hope to get a chance some time in the future to reproduce this problem in a simple example that I can share here.

The `decay` parameter to the tensorflow `assign_moving_average` function in this case is 0.99.

![batchnormweightexplosion](https://user-images.githubusercontent.com/18222703/44322682-51db8c00-a403-11e8-9d1b-f8554b626412.png)

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by MCMcCallum** [comment]
I guess these parameters need to be in the specific format provided. I've provided the parameters again below, adhering to the template format:

### System information

- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04 LTS
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: N/A
- **TensorFlow installed from (source or binary)**: Source
- **TensorFlow version (use command below)**: 1.8.0
- **Python version**: 3.5.2
- **Bazel version (if compiling from source)**: 0.11.1
- **GCC/Compiler version (if compiling from source)**: 4.9.4
- **CUDA/cuDNN version**: CUDA 9.1.85, CuDNN 7.1.2
- **GPU model and memory**: Gtx1070, 8GB memory
- **Exact command to reproduce**: N/A

**Comment by poxvoculi** [comment]
@brainnoise Is there enough information here to go on?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @poxvoculi: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @fchollet: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ymodak** [comment]
@MCMcCallum Is this still an issue for you? Can you please test with TF 1.13.rc0 and check? Thanks!

**Comment by ymodak** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by kwaegel** [comment]
@MCMcCallum Is your handcrafted update fix still working?

**Comment by blackPython** [comment]
I am facing the name problem. The moving_mean and moving_variance go to nan at training time and the effect of this can only be seen during the evaluation time. I am not using keras I am coding in tensorflow APIs and I am using tf.layers.batch_normalization. Is there a workaround?
